presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java

```diff
@@ -15,57 +15,142 @@
+         * theoretically incorrect thing to do.
+         *
+         * If you run into a case that you think justifies changing this, please consider
+         * that it's already pretty hard to determine if a failure is because the test
+         * is written incorrectly or because the actual plan really doesn't match a
+         * correctly written test. Having this throw makes it a lot easier to track down
+         * missing aliases in incorrect plans.
+         */
+        checkState(result != null, format("missing expression for alias %s", alias));
+        return result;
     }
 
-    public void updateAssignments(Map<Symbol, Expression> assignments)
+    private String toKey(String alias)
+    {
+        // Required because the SqlParser lower cases SymbolReferences in the expressions we parse with it.
+        return alias.toLowerCase();
+    }
+
+    private Map<String, Expression> getUpdatedAssignments(Map<Symbol, Expression> assignments)
     {
-        ImmutableMultimap.Builder<String, Expression> mapUpdate = ImmutableMultimap.builder();
+        ImmutableMap.Builder<String, Expression> mapUpdate = ImmutableMap.builder();
         for (Map.Entry<Symbol, Expression> assignment : assignments.entrySet()) {
-            for (String alias : map.keys()) {
-                if (map.get(alias).contains(assignment.getKey().toSymbolReference())) {
-                    mapUpdate.put(alias, assignment.getValue());
+            for (Map.Entry<String, Expression> existingAlias : map.entrySet()) {
+                if (assignment.getValue().equals(existingAlias.getValue())) {
+                    // Simple symbol rename
+                    mapUpdate.put(existingAlias.getKey(), assignment.getKey().toSymbolReference());
+                }
+                else if (assignment.getKey().toSymbolReference().equals(existingAlias.getValue())) {
+                    /*
+                     * Special case for nodes that we can alias symbols in the node's assignment map.
+                     * In this case, we've already added the alias in the map, but we won't include it
+                     * as a simple rename as covered above. Add the existing alias to the result if
+                     * the LHS of the assignment matches the symbol reference of the existing alias.
+                     *
+                     * This comes up when we alias expressions in project nodes for use further up the tree.
+                     * At the beginning for the function, map contains { NEW_ALIAS: SymbolReference("expr_2" }
+                     * and the assignments map contains { expr_2 := <some expression> }.
+                     */
+                    mapUpdate.put(existingAlias.getKey(), existingAlias.getValue());
                 }
             }
         }
-        map.putAll(mapUpdate.build());
+        return mapUpdate.build();
+    }
+
+    /*
+     * Update assigments in ExpressionAliases.map based on assignments given that
```

Comment:
This process could stand a high-level example (also downMatches and upMatches too, probably), it took me a while to figure this out and I was relying on the whiteboard session we had.